flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -197,13 +197,18 @@ BufferAndBacklog pollBuffer() {
 				return null;
 			}
 
+			boolean isExactlyOnceCheckpointBarrier = buffer.getDataType().isExactlyOnceCheckpointBarrier();
+			if (isExactlyOnceCheckpointBarrier) {
+				readView.blockOnCheckpointBarrier();
+			}
+
 			updateStatistics(buffer);
 			// Do not report last remaining buffer on buffers as available to read (assuming it's unfinished).
 			// It will be reported for reading either on flush or when the number of buffers in the queue
 			// will be 2 or more.
 			return new BufferAndBacklog(
 				buffer,
-				isAvailableUnsafe(),
+				isAvailableUnsafe() && !isExactlyOnceCheckpointBarrier,
```

Comment:
Remove `!isExactlyOnceCheckpointBarrier` to not mix this state with `ResultSubpartition` component, we can get this blocked state separately from `ResultSubpartitionView`, then every concept is clean.

Reply:
the downstream task of LocalChannel will check this ```isAvailable``` state and decide whether to add the corresponding input channel to available queue again. 

I will move this to constructor of BufferAndBacklog.

Reply:
I have unified this tag with isAvailable.